waspc/src/Analyzer/TypeChecker/Monad.hs

```diff
@@ -0,0 +1,43 @@
+{-# LANGUAGE TupleSections #-}
+
+module Analyzer.TypeChecker.Monad where
+
+import Analyzer.Parser.AST
+import Analyzer.Type
+import Analyzer.TypeChecker.TypeError
+import qualified Analyzer.TypeDefinitions as TD
+import Control.Monad.Except (Except, runExcept, throwError)
+import Control.Monad.Reader (ReaderT, asks, runReaderT)
+import Control.Monad.State (StateT, evalStateT, gets, modify)
+import qualified Data.HashMap.Strict as H
+
+type Bindings = H.HashMap Ident Type
+
+type T a = StateT Bindings (ReaderT TD.TypeDefinitions (Except TypeError)) a
```

Comment:
So we have State, Reader and Except all in one monad. Cool!

Reply:
I wonder if having short name as just `T` is good. I would probably name it `TypeChecker` without much thinking. `T` is short though. And in the context of TypeChecker I guess it is realtively clear what it is. But it looks more like a qualified import than name of the type. I would probably go with `TypeChecker`.